updating for 1.0.0

markcmiller86 opened the pull request:
In addition to updating for ZFP-1.0.0, I also did some other minor fixes

* Resolves #86 
* Resolves #81
* Resolves #77 
* Resolves #40 
* send stderr to `/dev/null` in GNU make grep in `config.make`
* define our own `ZFP_VERSION_STRING` if ZFP didn't define it (for older versions of ZFP)
* define our own `ZFP_CODEC` if ZFP didn't define it (for older versions of ZFP)
* switch to 4 digit hex version numbering for ZFP library
* capture `ZFP_CODEC` in HDF5 dataset `cd_vals` header
* Adjust logic for possible ZFP mismatch to use only codec number instead of lib version number
* minor test code improvements
   * disable CL args not appropriate on older versions of ZFP
   * disable read test features not appropriate
   * ensure `--help` CL arg to test codes works properly
* improved how test results are output
* enabled some tests that were not typically run before
* added some version compatibility tests

jwsblokland commented on test/test_write.c:
```diff
@@ -28,21 +28,30 @@ This file is part of H5Z-ZFP. Please also read the BSD license
+#error undefined ZFP library version symbols
 #endif
+
+#if !ZFP_VERSION_LE(0,9,9,9)
```
I think you accidentally removed the line "#if ZFP_HAS_CFP"

markcmiller86 replied:
Oh, sheesh. yes. Thanks. Will fix momentarily

jwsblokland commented on src/H5Zzfp.c:
```diff
@@ -48,9 +47,23 @@ and calls to bitstream methods with 'B ' as in
+#define ZFP_VERSION_NO                 ZFP_VERSION_NO_(ZFP_VERSION_MAJOR,ZFP_VERSION_MINOR,ZFP_VERSION_RELEASE,0)
+#elif defined(ZFP_VERSION_PATCH)
+#define ZFP_VERSION_NO                 ZFP_VERSION_NO_(ZFP_VERSION_MAJOR,ZFP_VERSION_MINOR,ZFP_VERSION_PATCH,0)
+#elif defined(ZFP_VERSION_PATCH)
```
I have the impression that this second "#elif defined(ZFP_VERSION_PATCH) will never be triggered.

markcmiller86 commented:
I need to add tests to read data from older H5Z-ZFP using 3 digit version number scheme.

markcmiller86 commented:
> I need to add tests to read data from older H5Z-ZFP using 3 digit version number scheme.

Ok, I've checked the code and there is only one place we examine the ZFP library version of the *stored* data and do something based on that...

https://github.com/LLNL/H5Z-ZFP/blob/3802c0f7406958008b628719ca2f523af14ff020/src/H5Zzfp.c#L508-L511

This logic is checking to see if the ZFP library version for the data stored in the file is *greater than* the ZFP library version it is currently compiled with. Although I switched to 4 hex digit version scheme, the logic is such that, for example, a stored value of `0x0055` (for older H5Z-ZFP and `0.5.5` ZFP) will *never* be greater than `0x0550` (newer H5Z-ZFP compiled with `0.5.5` ZFP).

That said, I am not entirely sure that older ZFP libary versions can't read data produced by newer versions and need to check with @lindstro on that question. @lindstro is there anything to worry about having *older* ZFP libs reading data produced by newer ZFP libs?

lindstro commented:
Keep in mind that `ZFP_VERSION` is based only on the library version.  For the _stored_ data, you'll most certainly want to use `ZFP_CODEC` instead, which is the version of the compressed format.  All library versions that read/write the same codec version are compatible.  Of course, if the H5Z-ZFP filter does not already use `ZFP_CODEC`, then you'd have to infer it from the library version.

lindstro commented:
A separate question has to do with backwards compatibility.  When `ZFP_CODEC` is incremented, we hope to still support reading files compressed using older codecs.  We've done a lot of ground work to allow this, including the possibility of linking to multiple instances of different versions of the zfp library, but that work is not quite mature yet.  Of course, we don't have to completely solve this problem until the next version of the `ZFP_CODEC`, which likely is years away.

markcmiller86 commented:
So, @lindstro, it sounds like this version comparison logic we have in H5Z-ZFP,

https://github.com/LLNL/H5Z-ZFP/blob/3802c0f7406958008b628719ca2f523af14ff020/src/H5Zzfp.c#L508-L511

really should NOT be there. It can wind up preventing a decompression that might otherwise have worked fine.

lindstro commented:
Yes, that looks like a problem to me.  Given how you're currently handling this, I think you need to hard code ranges of zfp library versions that are known to be compatible.

lindstro commented:
Although not officially exposed as part of the zfp API until version 0.5.1, `ZFP_CODEC` has implicitly matched `ZFP_VERSION_MINOR`, though zfp 1.0.0 departs from this convention.  `ZFP_CODEC=5` is the codec used in 0.5.x and 1.0.0.

As best I can tell, H5Z-ZFP was first released with zfp 0.5.0, so `ZFP_CODEC` has not yet changed. :-)

I suppose one could in principle build H5Z-ZFP with an even older version of zfp (e.g., 0.4.x), though I think H5Z-ZFP depends on the availability of zfp headers, which were not added until 0.5.0.  It makes sense to keep some kind of version check.  I think you now have all the necessary details to do this properly.

markcmiller86 commented:
I have tested this with ZFP versions 0.5.0, 0.5.2, 0.5.4, 0.5.5 and 1.0.0 installed either via vanilla make or via CMake. It seems to work. I wish I had an HDF5 dataset compressed with 0.4.0 so I could include a test case with mismatch in `ZFP_CODEC`. But, the codec mismatch logic here, 

https://github.com/LLNL/H5Z-ZFP/blob/c9bfcc09c30aad866d0f765c47bedfa8211c92b7/src/H5Zzfp.c#L501

seems to a) work and b) is sufficiently future proofed if the plugin is compiled against a newer ZFP library than we've currently tested.

lindstro commented:
Thanks for all the hard work, @markcmiller86.  You won't find an HDF5 data set compressed with 0.4.0 since H5Z-ZFP did not exist until 0.5.0. :-)  Nor could one be generated today since 0.4.0 does not support the zfp headers needed by H5Z-ZFP.  I think we're safe until codec 6 is released in the distant future.

lindstro commented on src/H5Zzfp.c:
```diff
@@ -480,6 +497,39 @@ get_zfp_info_from_cd_values(size_t cd_nelmts, unsigned int const *cd_values,
+       1.1.0 when ZFP was at version 1.0.0. So, anything we read from
+       the file that has two leading zeros needs to be shifted up 4 bits
+       to compare correctly */
+    if (zfpver_from_cd_val_data_in_file < 0x00FF)
```
Not that it matters in practice, but for correctness and readability, I think `0x00FF` ought to be `0x0100` the way the inequality is written.  Similarly with `ZFP_VERSION_NO < 0x04FF` below.

markcmiller86 commented:
@jwsblokland and @brtnfld I believe this is ready. I am hoping to tie push to `master` and release by COB Friday, Aug 12, 2022. Please have a look if you have a chance.

brtnfld reviewed (approved):
Looks good to me. I wonder if the ZFP_VERSION_LE should go in a header for the tests? We should ask if zfp could add these definitions. Providing similar definitions has come in handy for HDF5 users, @lindstro ?

lindstro commented:
I'd like to understand how `ZFP_VERSION_LE` differs from zfp's `ZFP_VERSION` and `ZFP_MAKE_VERSION` macros.  Though as discussed earlier, we should not be using those for H5Z-ZFP.  What matters is `ZFP_CODEC`.

markcmiller86 commented:
> Though as discussed earlier, we should not be using those for H5Z-ZFP. What matters is `ZFP_CODEC`.

It depends on whether you are talking about  info that gets persisted to an HDF5 file or just info that gets queried at compile time. We need  both and now use only `ZFP_CODEC` for persisting.

lindstro commented:
> > Though as discussed earlier, we should not be using those for H5Z-ZFP. What matters is `ZFP_CODEC`.
> 
> It depends on whether you are talking about info that gets persisted to an HDF5 file or just info that gets queried at compile time. We need both and now use only `ZFP_CODEC` for persisting.

Does this mean H5Z-ZFP files are no longer backwards compatible?

markcmiller86 commented:
> now use only `ZFP_CODEC` for persisting.

Well, we do indeed persist the ZFP lib version too so writers know what they used. But, we don't use that to control any logic in the filter...just informational in HDF5 meta data.

lindstro commented:
> > Does this mean H5Z-ZFP files are no longer backwards compatible?
> 
> No.

Is that a "no, it does not mean that", or "no, files written by H5Z-ZFP with zfp 1.0.0 cannot be read by older versions"?  I would hope that files written using any combination of zfp 0.5.x and 1.0.0 with any version of H5Z-ZFP would be readable by any other version.  But older versions of H5Z-ZFP did not embed `ZFP_CODEC`, which you're hinting is now being done, so does that not somehow break compatibility?

brtnfld commented:
Just to clarify, the HDF5 version macro is used to account for HDF5 API changes between versions, so it is a compile/feature macro. It has nothing to do with the HDF5 file format persisted to disk.

markcmiller86 commented:
> Looks good to me. I wonder if the ZFP_VERSION_LE should go in a header for the tests? We should ask if zfp could add these definitions. Providing similar definitions has come in handy for HDF5 users, @lindstro ?

So, I originally poured all the conditionally compiled version numbering logic into just the filter `.c` file because I didn't originally need it anywhere else.

But, then I needed to have similar logic in the test code too and wound up taking a different route there. Ideally, the test code should not use any knowledge from ZFP itself.  When I added the test for writing CFP compressed array via direct writes, the API for that changed in ZFP and the test won't compile against older ZFP libs without the conditionally checking the lib version.

All the other cases where I check ZFP version in the test code I can probably handle another way and wholly remove that logic.

Ultimately, the test code should really not use any information from ZFP headers. 

So, yes, I agree, it should be simplified/unified.

markcmiller86 commented:
> Is that a "no, it does not mean that", or "no, files written by H5Z-ZFP with zfp 1.0.0 cannot be read by older versions"? I would hope that files written using any combination of zfp 0.5.x and 1.0.0 with any version of H5Z-ZFP would be readable by any other version. But older versions of H5Z-ZFP did not embed `ZFP_CODEC`, which you're hinting is now being done, so does that not somehow break compatibility?

The first one...*no, it does not mean that*. If you look at this logic... https://github.com/LLNL/H5Z-ZFP/blob/7b19d42680ac59039a4b5fabc81bb172186fd332/src/H5Zzfp.c#L502-L533

You will note that for writers that used earlier versions of ZFP, we infer codec number from hand-coded logic that represents knowledge (which you explained to me above) about which ZFP versions use which codec. For this and newer writers, we used the codec info persisted to the file.

Likewise for readers (e.g. the ZFP lib the plugin is currently built with), we do a similar thing...use ZFP version number together with knowledge about which ZFP versions use which codec and otherwise use `ZFP_CODEC`. 

So, a file written two years ago with 0.5.5 does not have codec information persisted to the file. But, it does have ZFP library version information (as 3 digit hex). We get that (shift it up 4 bits because now we're using 4 hex digits for version number) and compare that to whatever ZFP we're compiled with.

I hope that makes sense but if not, have a closer look at the code. I think it achieves the desired goal.

jwsblokland reviewed (commented):
@markcmiller86 It all looks good to me. I did not expect that it would require so much changes. Have said that, I really like that you and @lindstro think ahead and try to implement a solution which is future proof.

One thing I did notice is that the h5repack test fails using either zfp 0.5.5 and 1.0.0. The H5Z-ZFP has been build using HDF5 1.12.1.

lindstro commented:
@markcmiller86 I think the way you're handling this make sense.  However, from a code readability standpoint, I want to understand what this line does:
```
     if (zfpver_from_cd_val_data_in_file < 0x0100) 
         zfpver_from_cd_val_data_in_file <<= 4; 
```
Is this querying if the zfp library version is pre 1.0.0?  Wouldn't that more accurately be expressed as `< 0x1000`?  Or does `< 0x0100` mean zfp pre 1.0.0 written with "old" H5Z-ZFP, with the range 0x0100 to 0x0fff reserved for zfp pre 1.0.0 with "new" H5Z-ZFP, where the third hex digit from the right is `ZFP_CODEC`?  Where does `ZFP_CODEC` get stored with zfp 1.0.0?

markcmiller86 commented:
> One thing I did experience is that the h5repack test fails using either zfp 0.5.5 and 1.0.0. The used H5Z-ZFP filter has been build using HDF5 1.12.1.

Hmm...did you test that before or after [db519b3](https://github.com/LLNL/H5Z-ZFP/pull/85/commits/db519b352830a2741f417b34c9ca6cfe6055fde1)? I did find and fix an issue with it and you can see it is now...

![Screen Shot 2022-08-12 at 8 47 01 AM](https://user-images.githubusercontent.com/5720676/184393384-ff9beb8d-741a-401c-acf9-8236a86dd19e.png)

markcmiller86 commented:
@jwsblokland  I did switch the h5repack test logic from using `dc` to `perl` to compute a float division and round it. Maybe that is the source of problems for you?

markcmiller86 commented:
> Is this querying if the zfp library version is pre 1.0.0?

Not really. Its quering if the zfp library version stored in the file has two leading zeros. Prior to ZFP version 1.0.0 (and H5Z-ZFP version 1.1.0 which we are getting ready to release here now), two things were true...

- three hex digit version numbers were being used
- the major number for ZFP was always zero

Combined, that means that any data written previously has two leading zeros in the ZFP version stored to the file. This is testing for that and then adjusts/corrects it by shifting up 4 bits.

Any writers using an old ZFP lib but with this new H5Z-ZFP filter will, by virtue of the 4 digit numbering scheme it uses have at least one of the two leading digits non-zero.

Any writers using new ZFP lib but with old H5Z-ZFP filter would have...oh, wait...I think my logic isn't handling that case quite correctly.

markcmiller86 commented:
> I did not expect that it would require so much changes.

Handling ZFP-1.0.0 didn't really take too much.

Because I don't often delve into this code, while I am here, I wanna address several other things on the *to do* list. Honestly, from a software quality engineering standpoint, its probably very bad form to do this all in one PR if anyone strongly wishes I wouldn't do that, I can close this one and re-submit several smaller PRs. Its just more overhead for me and anyone else I enlist to review 😉

vasole commented:
As last option, there is always the alternative to use a different HDF5 filter ID...

jwsblokland commented:
> @jwsblokland  I did switch the h5repack test logic from using `dc` to `perl` to compute a float division and round it. Maybe that is the source of problems for you?

Next week, I will try to figure out what the issue of my failed test is. One thing I already noticed is that the repacked file was larger compared to the original one. This is already weird, I think. Therefore, I assume I did something wrong.

jwsblokland commented on test/Makefile:
```diff
@@ -254,10 +327,17 @@ test-reversible: plugin test_write_plugin
+	@echo " "; \
+	pad=$$(printf '%*s' "$(padlimit)"); \
+	pad=$${pad// /.}; \
+	x="h5repack -f UD=1,32013,6,3,0,3539053052,1062232653,0,0"; \
```
If I understand the help of h5repack application correctly, it should be specified as
```
h5repack -f UD=32013,0,6,3,0,3539053052,1062232653,0,0
```
At least, if we want to make use of the H5Z-ZFP filter during the repacking of the HDF5 file. However, I use this line I got the HDF5 error
```
HDF5-DIAG: Error detected in HDF5 (1.12.1) thread 0:
  #000: /glb/data/SeismicApps3/nljbl6/HDF5.custom/1.12.1_patch/source/src/H5T.c line 2090 in H5Tget_class(): not a datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
```
Currently, I have no idea what the reason for this error is.

Note that in the original specification we made accidentally use of the GZIP filter.

brtnfld replied:
Correct, the filter ID should be first. Are you using a version of h5repack that has HDF5 statically linked in? HDF5 using dynamic plugins use the dlopen call, which must use shared libraries. Could you see if you have _h5repack-shared_?

markcmiller86 replied:
Ok, I think I know what I did...I used default paths to `h5repack` and am getting `1.10.0-patch` as the version. I think that still required the leading `1` or `0` to indicate mandatory or optional filter. Let me go check which version our CI is using.

markcmiller86 commented:
@jwsblokland ok thanks for persisting :muscle:. I was confused by accidentally using default path to `h5repack` in my shell and then incorrect logic for assessing size of repacked file (to confirm the filter had taken some action) and wound up breaking the `h5repack` tests and docs (the had effectively already been fixed for newer HDF5's and I hadn't paid attention to that too). 

The filter ID comes *first* followed by either a `0` meaning the filter is *mandatory* or a `1` meaning the filter is optional followed by the raw CD vals data, comma separated with no spaces.

lindstro commented:
> > Is this querying if the zfp library version is pre 1.0.0?
> 
> Not really. Its quering if the zfp library version stored in the file has two leading zeros. Prior to ZFP version 1.0.0 (and H5Z-ZFP version 1.1.0 which we are getting ready to release here now), two things were true...
> 
> * three hex digit version numbers were being used
> * the major number for ZFP was always zero

I believe this all to be true, but I'd like to understand better the strategy to handle various combinations of zfp and H5Z-ZFP library versions and what is stored in `cd_vals`.  Please correct me where I'm wrong:

- zfp <= 0.5.5 and H5Z-ZFP <= 1.0.0, `cd_vals` version = 0.0.`ZFP_VERSION_MAJOR`.`ZFP_VERSION_MINOR`.
- zfp <= 0.5.5 and HZ5-ZFP >= 1.1.0, `cd_vals` version = 0.`ZFP_CODEC`.`ZFP_VERSION_MAJOR`.`ZFP_VERSION_MINOR`.
- zfp >= 1.0.0 and HZ5-ZFP >= 1.1.0, `cd_vals` version = `ZFP_VERSION_MAJOR`.`ZFP_VERSION_MINOR`.`ZFP_VERSION_PATCH`.? (`ZFP_CODEC`, perhaps?).
- zfp >= 1.0.0 and HZ5-ZFP <= 1.1.0, `cd_vals` version = ? (I guess this isn't handled yet).

markcmiller86 commented:
@lindstro for all versions of H5Z-ZFP prior to this PR, the ZFP *library* version number was defined in the CPP symbol `ZFP_VERSION_NO` which mapped to `0x0 Maj Min Rel` (with always a leadig zero in the most significant hex digit). So, for example, ZFP library version `0.5.0` lead to `ZFP_VERSION_NO` with value`0x0050` and ZFP library version `0.5.2` led to `0x0052`.

For all versions of H5Z-ZFP prior to this PR, `ZFP_VERSION_NO` was then captured in `cd_vals` dataset header in an HDF5 file as the upper 16 bits of `cd_vals[0]`. The lower 16 bits of `cd_vals[0]` captured `H5Z_FILTER_ZFP_VERSION_NO` which also used a 3-digit scheme with a leading `0x0`.

With this PR, we have changed `ZFP_VERSION_NO` to use a 4 hex digit scheme (so no automatic leading zero in the form of a `0x0` being pasted with other `Maj`, `Min`, `Pat` (and `Tweak`) values. So, now, ZFP library version `0.5.0` leads to `ZFP_VERSION_NO` of `0x0500` and `0.5.2` to `0x0520`. ZFP version `1.0.0` (with `tweak==0` also) leads to `0x1000`.

In addition, with this PR we are also now storing `ZFP_CODEC` to the file (were using the upper 4 bits of the lower 16-bit word of `cd_vals[0]` (which was and still is always zero for `H5Z_FILTER_ZFP_VERSION_NO` as you can see here https://github.com/LLNL/H5Z-ZFP/blob/843f94082500ece65f5f25c5eabadf6ad39a38bb/src/H5Zzfp.c#L319

If `ZFP_CODEC` is not defined for the current version of ZFP library we are building with, we define it to be the same as `ZFP_VERSION_MINOR` as in https://github.com/LLNL/H5Z-ZFP/blob/843f94082500ece65f5f25c5eabadf6ad39a38bb/src/H5Zzfp.c#L70-L72.

We removed logic that compared ZFP library versions and instead now do a codec version mismatch check in the decompression part of the filter...https://github.com/LLNL/H5Z-ZFP/blob/843f94082500ece65f5f25c5eabadf6ad39a38bb/src/H5Zzfp.c#L581-L582

We attempt to decode the args to this call from `cd_vals` we've read from the file here, https://github.com/LLNL/H5Z-ZFP/blob/843f94082500ece65f5f25c5eabadf6ad39a38bb/src/H5Zzfp.c#L562-L564

And, then we do our mismatch logic here...

https://github.com/LLNL/H5Z-ZFP/blob/843f94082500ece65f5f25c5eabadf6ad39a38bb/src/H5Zzfp.c#L503-L552

Now, to your questions...



> - zfp <= 0.5.5 and H5Z-ZFP <= 1.0.0, `cd_vals` version = 0.0.`ZFP_VERSION_MAJOR`.`ZFP_VERSION_MINOR`.

Well, not quite....version= 0x0 # `ZFP_VERSION_MAJOR` # `ZFP_VERSION_MINOR` # `ZFP_VERSION_RELEASE` where `#` is CPP token pasting operator

> - zfp <= 0.5.5 and HZ5-ZFP >= 1.1.0, `cd_vals` version = 0.`ZFP_CODEC`.`ZFP_VERSION_MAJOR`.`ZFP_VERSION_MINOR`.

No. version=`0x` # `ZFP_VERSION_MAJOR` # `ZFP_VERSION_MINOR` # `ZFP_VERSION_PATCH` # `ZFP_VERSION_TWEAK` where `#` is CPP token pasting operator and codec handled as described above.

> - zfp >= 1.0.0 and HZ5-ZFP >= 1.1.0, `cd_vals` version = `ZFP_VERSION_MAJOR`.`ZFP_VERSION_MINOR`.`ZFP_VERSION_PATCH`.? (`ZFP_CODEC`, perhaps?).

No. same as case immediately above.

> - zfp >= 1.0.0 and HZ5-ZFP <= 1.1.0, `cd_vals` version = ? (I guess this isn't handled yet).

Right, old writers didn't explicitly capture codec. They captured ZFP library version and we don't know how future versions of ZFP will map to codec versions. So, if someone uses an older filter with newer ZFP library, we have no idea what codec they used. But, we can adjust for this whenever the codec number is incrimented and will otherwise, for now anyways, have to guess its `5`.

jwsblokland commented:
@markcmiller86 and @brtnfld I got all the tests working and passing. For the h5repack test I needed to make use of h5repack-shared. Thanks for reminding me that for dynamic plugins you need to make use of the shared HDF5 library. Now, I am wondering if it would be possible to make use of the correct version of ``h5repack``. I mean for HDF5 1.12 you need to make use of ``h5repack-shared`` while for HDF5 1.10 it is ``h5repack``. At least that is my impression. Or is it related to the fact I have used CMake to build HDF5 1.12 library.

Furthermore, I noticed that the return code of a failed has been changed from 1 to 0. I am not sure if this is a good idea. Now, if one of more tests fail it looks like everything is ok except for the error message in the console output.

markcmiller86 commented:
> @markcmiller86 and @brtnfld I got all the tests working and passing. For the h5repack test I needed to make use of h5repack-shared. Thanks for reminding me that for dynamic plugins you need to make use of the shared HDF5 library. Now, I am wondering if it would be possible to make use of the correct version of `h5repack`. I mean for HDF5 1.12 you need to make use of `h5repack-shared` while for HDF5 1.10 it is `h5repack`. At least that is my impression. Or is it related to the fact I have used CMake to build HDF5 1.12 library.

I think hdf5 utilities with `-shared` in the name has to do with static builds of HDF5. We can't really test *as plugin* mode of this filter against a statically installed HDF5. We can only test the *as library* mode. We could add logic to check for that I guess. There are `h5diff` checks that we do too and those tests will not work either.

> 
> Furthermore, I noticed that the return code of a failed has been changed from 1 to 0. I am not sure if this is a good idea. Now, if one of more tests fail it looks like everything is ok except for the error message in the console output.

Can you elaborate? Which test? I adjusted `Makefile` logic to `exit 0` from subshells after reporting pass/fail result of any test so that it would continue on to run other targets/tests for the `check` target instead of exiting immediately and not running any further tests. But, that does mean the exit code for the `make check` command may be zero even if some test(s) failed. I can fix that but maybe not in a way that works with `make -j <some large number>`.

markcmiller86 commented:
> There are `h5diff` checks that we do too and those tests will not work either.

It doesn't look like `h5ls` returns non-zero exit status if it is unable to load a plugin needed to display a dataset's contents. I suspect this is true across the board with HDF5 tools...they won't necessarily produce non-zero exit status in case they are unable to find/load a plugin needed to read a given dataset. That means, logic for checking failure to employ a tool like `h5repack` or `h5diff` in service of confirming a given test here has worked as expected or not is more complex. I already do not like complexity of `test/Makefile` (i've added a ton of shell code to it recently which would be better left to CTest in the future I think).

jwsblokland commented:
> Can you elaborate? Which test? I adjusted Makefile logic to exit 0 from subshells after reporting pass/fail result of any test so that it would continue on to run other targets/tests for the check target instead of exiting immediately and not running any further tests. But, that does mean the exit code for the make check command may be zero even if some test(s) failed. I can fix that but maybe not in a way that works with make -j <some large number>.

That is exactly what I observe. In the past, when the h5repack failed I got an error message together with a non-zero return value. For me, a clear indication that something is wrong. Now, if the h5repack test fails it more difficult to see it had failed. Fortunately, the number of tests is limited so you can still notice it in the console output. Furthermore, I tried for example ``make -j2 ... check`` which fails. So it looks like parallel testing is not possible at this moment.

I agree with you we should convert all existing tests to CTest. It should not be so difficult but it requires significant amount of time. Having said this all, I am happy that you have implement support for the latest ZFP version.

markcmiller86 commented:
Ok, so next step is Spack recipe. I've been working on that but not with much luck. I'll merge this and prepare are release now.
